src/pylibsshext/session.pyx

```diff
@@ -219,36 +219,59 @@ cdef class Session(object):
                 libssh.ssh_disconnect(self._libssh_session)
                 raise
 
-        if kwargs.get('private_key'):
+        # We need to userauth_none before we can query the available auth types
+        rc = libssh.ssh_userauth_none(self._libssh_session, NULL)
+        if rc == libssh.SSH_AUTH_SUCCESS:
+            # Huh, it worked?
+            return
+        if rc == libssh.SSH_AUTH_ERROR:
+            raise LibsshSessionException("Error while fetching list of supported authentication methods")
+
+        supported_auth = libssh.ssh_userauth_list(self._libssh_session, NULL)
+
+        if kwargs.get('private_key') and supported_auth & libssh.SSH_AUTH_METHOD_PUBLICKEY:
             # try authenticating with a given private key
             try:
                 self.authenticate_specific_pubkey(
                     kwargs['private_key'],
                     kwargs.get('private_key_password'),
                 )
-                return
             except LibsshSessionException as ex:
                 saved_exception = ex
+            else:
+                return
 
         # try authenticating with a password
-        if kwargs.get('password'):
+        if kwargs.get('password') and supported_auth & libssh.SSH_AUTH_METHOD_PASSWORD:
             try:
                 self.authenticate_password(kwargs["password"])
+            except LibsshSessionException as ex:
+                saved_exception = ex
+            else:
                 return
+
+        # try authenticating with keyboard-interactive (without the interactive)
```

Comment:
What do you mean by "without the interactive"?

Reply:
I suppose "without the keyboard" would be more accurate, just trying to convey that we aren't actually expecting there to be a user interactive element as one might expect.

Reply:
maybe write that down? "emulate a keyboard input without requesting it from the user"